client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -81,11 +77,22 @@ private final class PoolManager[F[_], A <: Connection[F]](
           IO(callback(Left(error)))
       }
     } else {
-      val message =
-        s"Invariant broken in ${this.getClass.getSimpleName}! Tried to create more connections than allowed: ${stats}"
-      val error = new Exception(message)
-      logger.error(error)(message)
-      callback(Left(error))
+      addToWaitQueue(key, callback)
+    }
+
+  private def getNonEmptyKeys: List[RequestKey] =
+    idleQueues.foldLeft(List.empty[RequestKey]) {
+      case (l, (k, q)) =>
+        if (q.nonEmpty) l :+ k
+        else l
+    }
```

Comment:
Wait, aren't they all non-empty?

Reply:
That's what I thought, and I was yelled at by the test which showed a scenario where this might not be the case.

Reply:
Wait, I forgot how this made sense !!

Reply:
Yeah, if none of the connections are idle because of slow requests, we would get a situation where maxTotal has been reached but idleQueues are empty.

Reply:
But don't we remove the idleQueue from the map when we dequeue the last idle connection?

Reply:
Yes, but there were no idle connections to begin with, so there was no entry.